tony-portal/app/history/HistoryFileCleaner.java

```diff
@@ -0,0 +1,110 @@
+package history;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.linkedin.tony.TonyConfigurationKeys;
+import com.typesafe.config.Config;
+import hadoop.Requirements;
+import java.io.IOException;
+import java.time.Instant;
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.time.temporal.ChronoUnit;
+import java.time.temporal.TemporalAdjusters;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import play.Logger;
+import utils.ConfigUtils;
+
+
+@Singleton
+public class HistoryFileCleaner {
+  private static final Logger.ALogger LOG = Logger.of(HistoryFileCleaner.class);
+
+  @Inject
+  public HistoryFileCleaner(Config appConf, Requirements requirements) {
+    FileSystem fs = requirements.getFileSystem();
+    Path intermediateDir = requirements.getIntermediateDir();
+    Path finishedDir = requirements.getFinishedDir();
+    long retentionSec = ConfigUtils.fetchIntConfigIfExists(appConf,
+        TonyConfigurationKeys.TONY_HISTORY_RETENTION_SECONDS,
+        TonyConfigurationKeys.DEFAULT_TONY_HISTORY_RETENTION_SECONDS);
+
+    ScheduledExecutorService scheduledThreadPool = Executors.newScheduledThreadPool(1);
+    long cleanerIntervalMs = ConfigUtils.fetchIntConfigIfExists(appConf,
+        TonyConfigurationKeys.TONY_HISTORY_CLEANER_INTERVAL_MS,
+        TonyConfigurationKeys.DEFAULT_TONY_HISTORY_CLEANER_INTERVAL_MS);
+
+    LOG.info("Retention period is " + retentionSec + " seconds");
+    LOG.info("Starting background history file cleaner thread, will run every " + cleanerIntervalMs + " milliseconds.");
+    scheduledThreadPool.scheduleAtFixedRate(() -> {
+      LocalDate cutOffDate = LocalDateTime.now().minusSeconds(retentionSec).toLocalDate();
+      LOG.info("Removing all history files older than " + cutOffDate);
+      try {
+        cleanFinishedDir(fs, finishedDir, cutOffDate);
+        cleanIntermediateDir(fs, intermediateDir, cutOffDate);
+      } catch (Exception e) {
+        LOG.error("Encountered exception while cleaning history directories", e);
+      }
+    }, 0, cleanerIntervalMs, TimeUnit.MILLISECONDS);
+  }
+
+  /**
+   * Deletes all year/month/day directories in the finished dir prior to the cutoff date.
+   */
+  @VisibleForTesting
+  static void cleanFinishedDir(FileSystem fs, Path finishedDir, LocalDate cutOffDate) throws IOException {
```

Comment:
why is cleanFinishedDir so different from cleanIntermediateDir ?

Reply:
The structure of finished is yyyy/mm/dd/<app_id> whereas intermediate directly contains <app_id> folders.

For purging finished, we have to parse yyyy/mm/dd and interpret as a date, whereas for intermediate, I'm just using the last modification time of the <app_id> directory as a heuristic.